Suppress Nurbs warnings

SergioRAgostinho opened the pull request:
While preparing to work on the release items, I noticed nurbs was spewing out an impressive number of warnings. I decided to finally clean it up. There's only a deprecation warning missing.

Situations in which I commented variables instead of simply deleting them are usually a sign that there is commented code in the function's body which originally intended to use them. 

Most if not all, ctor base class initializations that I added to suppress the compiler warnings are actually no-ops. Probably the reason they were not being invoked in the first place.

taketwo commented on surface/src/3rdparty/opennurbs/opennurbs_bezier.cpp:
```diff
@@ -1374,7 +1374,7 @@ bool ON_BezierCurve::GetCV( int i, ON::point_style style, double* Point ) const
   case ON::euclidean_rational:
     Point[dim] = w;
-    // no break here
+    // fall through
```
We've merged a new `FALLTHROUGH` macro some days ago: https://github.com/PointCloudLibrary/pcl/pull/3320/commits/df40206d46205879caadc8f253c01db779bb6bc5, I think it will fit here nicely?

SergioRAgostinho commented:
Sorry for the early force push. I forgot about it.

taketwo commented:
There are a few more `sign-compare` warnings left in Nurbs, see lines 3716-3806 in [Ubuntu CI log](https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=6495&view=logs&j=377bee81-b530-5eab-ac40-709ee4d08646&t=4d16a4ce-53cb-56fa-8e89-15ea4353307f&l=3716), would you mind adding a commit for them as well?

SergioRAgostinho commented:
Small heads-up. I found it strange that I was not seeing those sign-compare warnings during my build. I'm using g++ 7.4 on a Ubuntu distro. I double checked my `CMAKE_CXX_FLAGS` and they're empty. I'm gonna dig around this new issue before proceeding with the last commit on this PR.

SergioRAgostinho commented:
The ci job reports using gcc 5.4.
``` 
-- The C compiler identification is GNU 5.4.0
-- The CXX compiler identification is GNU 5.4.0
``` 
I manually compiled gcc 5.5.0 (could not get 5.4.0 to compile) and gave it a try at compiling `pcl_surface`and it's not outputting the sign-compare warnings.

Is the build outputting anything in your environment?

taketwo commented:
I have GCC 5.5 on my Ubuntu 16 machine and it also does not output sign warnings.

I'm puzzled why did CI job report GCC 5.4 though. It's running on a Ubuntu 16 based container which I re-created just a few days ago, so it should also have GCC 5.5. I will investigate tomorrow.

By the way, here is how you can quickly reproduce a CI run using docker. First, get the image:

```bash
docker pull pointcloudlibrary/env:16.04
```

Second, run the image mapping PCL source directory into it. Assuming you are in the repository root:

```bash
docker run --rm -v $PWD:/code -it pointcloudlibrary/env:16.04 /bin/bash
```

This will get you into a Docker container. Create a build directory:

```bash
mkdir /tmp/build && cd /tmp/build
```

Configure and make:

```bash
cmake /code && make
```

taketwo commented:
I implemented the changes needed to suppress `-Wsign-compare` and took the liberty to commit them to your branch. Ubuntu 16.04 job on CI runs clean from NURBS warnings, so I think this is ready to be merged.

kunaltyagi commented:
> Also starting from a fresh ubuntu:16.04 container and installing gcc gives me version 5.4

It might be because the image is cached and you need to pull the base image too. Use `--pull` in your `docker build` command to make docker check for a newer base image. (I'm sorry, I haven't built the docker image myself to confirm my hunch)

taketwo commented:
I tried to `docker build --pull` the Dockerfile that we have in the repository, but the compiler is still the same, GCC 5.4.0.

kunaltyagi commented on surface/src/on_nurbs/fitting_curve_2d_apdm.cpp:
```diff
@@ -1046,7 +1046,7 @@ FittingCurve2dAPDM::findClosestElementMidPoint (const ON_NurbsCurve &nurbs, cons
     double dxi = xi1 - xi0;
 
-    for (unsigned j = 0; j < nurbs.Order (); j++)
+    for (unsigned j = 0; j < static_cast<unsigned int>(nurbs.Order ()); j++)
```
The casts are all over the place. (to unsigned int, size_t). Maybe only one is required and not all (for `nurbs.Order()`) variants

kunaltyagi commented on surface/src/on_nurbs/fitting_curve_2d_asdm.cpp:
```diff
@@ -289,7 +289,7 @@ FittingCurve2dASDM::assembleInterior (double wInt, double sigma2, double rScale,
     Eigen::Vector2d pt, t, n;
     double error;
-    if (p < int (m_data->interior_param.size ()))
+    if (p < static_cast<unsigned int>(m_data->interior_param.size ()))
```
`p` as `std::size_t` negates the need for cast, esp since other comparisons with `XYZ.size()` are uncasted

SergioRAgostinho commented:
Sorry for the absence. In situations like this feel free both of you to push stuff to my PR branches whenever required.

SergioRAgostinho commented:
Oops. Forgot about addressing your comments. The changes LGTM. 

I would personally not deprecate or make any other changes considered to be stricly required to  3rd party code but there's indeed the "risk" someone is using directly the OpenNurbs version we're packaging.

SergioRAgostinho commented:
I just rebased this branch on top the current master and there still a number of `size_t` occurrences present in the code. :confused:
